### PR TITLE
changes to reduce thread churn and creation

### PR DESCRIPTION
**Pull request by gjdeval**
Changes to reduce thread churn and creation:
1) tamp down thread creation in no/low-load scenario
2) require a stronger signal from historical data to increase/decrease pool size
3) fix an NPE found recently in setPoolIncrementDecrement

### CODE REVIEW

File: dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java [hunk 1]
```diff
@@ -1266,21 +1318,34 @@ synchronized String evaluateInterval() {
                 cpuHigh = true;
             }
 
+            boolean lowTput = false;
+            if (throughput / poolSize < lowTputThreadsRatio && queueDepth == 0) {
```
Comment (jhanders34) [inline dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java]:
```suggestion
            if (queueDepth == 0 && throughput / poolSize < lowTputThreadsRatio) {
```
It makes sense to do the cheaper conditional first.  No reason to do complicated division if queueDepth is not 0.

File: dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java [hunk 2]
```diff
@@ -1266,21 +1318,34 @@ synchronized String evaluateInterval() {
...
-            double growScore = getGrowScore(poolSize, forecast, throughput, queueDepth, cpuHigh);
+            double shrinkScore = getShrinkScore(poolSize, forecast, throughput, queueDepth, cpuHigh, lowTput);
+            double growScore = getGrowScore(poolSize, forecast, throughput, queueDepth, cpuHigh, lowTput);
 
+            // Adjust the poolsize only if one of the scores is both larger than the scoreFilterLevel
+            // and sufficiently larger than the other score. These conditions reduce poolsize fluctuation
+            // which might arise due to a weak or noisy signal from the historical throughput data.
             int poolAdjustment = 0;
-            if (growScore > shrinkScore) {
+            if (growScore > growScoreFilterLevel && growScore > (shrinkScore + growShrinkDiffFilter)) {
```
Comment (jhanders34) [inline dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java]: This doesn't look right.  I am thinking something more like:
```
if (Math.abs(growScore - shrinkScore) > growShrinkDiffFilter) {
    if (growScore > growScoreFilterLevel && growScore > shrinkScore) {
        poolAdjustment = poolIncrement;
    } else if (shrinkScore > shrinkScoreFilterLevel && shrinkScore > growScore) {
        poolAdjustment = -poolIncrement;
    }
}
```

File: dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java (hunk 1 above)
Comment (gjdeval) [reply dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java]: Agreed, changed.

File: dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java (hunk 2 above)
Comment (gjdeval) [reply dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java]: This is I think functionally equivalent to the prior code, but will in some cases be more efficient, so I agree and changed accordingly.

**Comment by gjdeval** [comment]
#build

**Comment by gjdeval** [comment]
Fixes #5779

**Pull request merged**